MultiBuild/BlueprintManager.cs

```diff
@@ -554,124 +523,67 @@ public static List<BuildPreview> Paste(Vector3 targetPos, float yaw, bool pasteI
-                bp.desc = desc;
-                bp.item = building.itemProto;
-                bp.recipeId = building.recipeId;
-                bp.lpos = absoluteBuildingPos;
-                bp.lrot = absoluteBuildingRot;
-
-                Pose pose = new Pose(absoluteBuildingPos, absoluteBuildingRot);
-
-                int objId = InserterPoses.AddOverride(pose, building.itemProto);
-
-                pastedEntities.Add(building.originalId, new PastedEntity()
-                {
-                    type = EPastedType.building,
-                    index = i,
-                    sourceBuilding = building,
-                    pose = pose,
-                    objId = objId,
-                    buildPreview = bp
-                });
-                absolutePositions.Add(absoluteBuildingPos);
-                previews.Add(bp);
+                PasteBuilding(i, data.copiedBuildings[i], targetSpr, yaw);
             }
 
-
             for (int i = 0; i < data.copiedBelts.Count; i++)
             {
-                BeltCopy belt = data.copiedBelts[i];
-                Vector2 newRelative = belt.cursorRelativePos.Rotate(yawRad, belt.originalSegmentCount);
-                Vector2 sprPos = newRelative + targetSpr;
-
-                float rawLatitudeIndex = (sprPos.x - Mathf.PI / 2) / 6.2831855f * 200;
-                int latitudeIndex = Mathf.FloorToInt(Mathf.Max(0f, Mathf.Abs(rawLatitudeIndex) - 0.1f));
-                int newSegmentCount = PlanetGrid.DetermineLongitudeSegmentCount(latitudeIndex, 200);
-
-                float sizeDeviation = belt.originalSegmentCount / (float)newSegmentCount;
-                if (sizeDeviation > currentMaxWidth)
-                    currentMaxWidth = sizeDeviation;
-
-                if (useExperimentalWidthFix && sizeDeviation < lastMaxWidth)
-                    sizeDeviation = lastMaxWidth;
-
-                sprPos = new Vector2(newRelative.x, newRelative.y * sizeDeviation) + targetSpr;
-
-                Vector3 absoluteBeltPos = sprPos.SnapToGrid(belt.altitude * 1.3333333f / 2);
-
-                // belts have always 0 yaw
-                Quaternion absoluteBeltRot = Maths.SphericalRotation(absoluteBeltPos, 0f);
-
-                BuildPreview bp = BuildPreview.CreateSingle(belt.itemProto, belt.itemProto.prefabDesc, true);
-                bp.ResetInfos();
-                bp.desc = belt.itemProto.prefabDesc;
-                bp.item = belt.itemProto;
-
-                bp.lpos = absoluteBeltPos;
-                bp.lrot = absoluteBeltRot;
-                bp.outputToSlot = -1;
-                bp.outputFromSlot = 0;
-
-                bp.inputFromSlot = -1;
-                bp.inputToSlot = 1;
-
-                bp.outputOffset = 0;
-                bp.inputOffset = 0;
+                PasteBelt(i, data.copiedBelts[i], targetSpr, yaw);
+            }
 
-                Pose pose = new Pose(absoluteBeltPos, absoluteBeltRot);
+            if (pasteInserters && data.copiedInserters.Count > 16)
```

Comment:
The ThreadPool should handle scalability by itself, I'd recommend having `Environment.ProcessorCount - 1` work items queued. (leave the UI thread on its own thread).

That way, you don't force dual core processors to context switch all the time.